x/cash/dynamicfee_test.go

```diff
@@ -0,0 +1,240 @@
+package cash
+
+import (
+	"testing"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/store"
+	"github.com/iov-one/weave/x"
+)
+
+func TestDynamicFeeDecorator(t *testing.T) {
+	perm1 := weave.NewCondition("sigs", "ed25519", []byte{1, 2, 3})
+	//perm2 := weave.NewCondition("sigs", "ed25519", []byte{3, 4, 5})
+	perm3 := weave.NewCondition("custom", "type", []byte{0xAB})
+
+	collectorAddr := perm3.Address()
+
+	coinp := func(w, f int64, t string) *x.Coin {
+		c := x.NewCoin(w, f, t)
+		return &c
+	}
+
+	walletObj := func(a weave.Address, w, f int64, ticker string) orm.Object {
+		t.Helper()
+		obj, err := WalletWith(a, coinp(w, f, ticker))
+		if err != nil {
+			t.Fatalf("cannot create a wallet: %s", err)
+		}
+		return obj
+	}
+
+	cases := map[string]struct {
+		signers    []weave.Condition
+		handler    *handlerMock
+		minumumFee x.Coin
+		txFee      x.Coin
+		// Wallet state created before running Check
+		initWallets []orm.Object
+		// Wallet state applied after running Check but before running Deliver
+		updateWallets []orm.Object
+
+		wantCheckErr     error
+		wantCheckTxFee   x.Coin
+		wantDeliverErr   error
+		wantDeliverTxFee x.Coin
+		wantGasPayment   int64
+	}{
+		"on success full transaction fee is charged": {
+			signers: []weave.Condition{perm1},
+			handler: &handlerMock{},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 1, 0, "BTC"),
+			},
+			minumumFee:       x.NewCoin(0, 23, "BTC"),
+			txFee:            x.NewCoin(0, 421, "BTC"),
+			wantCheckTxFee:   x.NewCoin(0, 421, "BTC"),
+			wantDeliverTxFee: x.NewCoin(0, 842, "BTC"),
```

Comment:
I hope this is correct, that after successful `Deliver` call, we charged the transaction fee twice (421 + 421 = 842).

Reply:
This is not correct. I was reminded about this recently that the check state is not persisted. See https://github.com/iov-one/weave/blob/master/app/commit.go#L44 Therefore you should reset the store before passing it to `deliver` in your tests.